feat(containerdisk): Add support for Debian containerdisk and sha512 checksums

jcanocan opened the pull request:
**What this PR does / why we need it**:

It adds support for sha512 checksums, which are required by Debian. Debian image repository only provides sha512 checksums. Moreover, it adds support for Debian 11 and 12 containerdisk on x86_64 and aarm64.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes # [CNV-51712](https://issues.redhat.com/browse/CNV-51712)

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Added support for Debian 11 and 12 containerdisks.
```

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,117 @@
+		baseURL = fmt.Sprintf(baseURLFmt, d.VersionName+"/daily")
+		qCowImageName = fmt.Sprintf(baseQcow2Fmt, d.Version, architecture.GetImageArchitecture(d.Arch)+"-daily")
+	} else {
+		panic(fmt.Sprintf("can't understand provided version %s", d.Version))
```
Please don't panic, you can return an error instead.

jcanocan replied:
Done.

0xFelix commented on artifacts/centosstream/centos-stream.go:
```diff
@@ -91,7 +91,7 @@ func (c *centos) Inspect() (*api.ArtifactDetails, error) {
 	if checksum, exists := checksums[candidate]; exists {
 		return &api.ArtifactDetails{
-			SHA256Sum:            checksum,
+			SHASum:               checksum,
```
Rename to `Checksum`?

jcanocan replied:
IMHO, we are currently just supporting SHA digests. Therefore, it looks appropriated as it is. Maybe it can be changed as `SHAChecksum`. WDYT?

jcanocan replied:
Done.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,117 @@
+		Name:            "debian",
+		Version:         d.Version,
+		Description:     description,
+		ExampleUserData: *d.ExampleUserData,
```
`EnvVariables` and `Arch` need to be wired up?

jcanocan replied:
Done.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,117 @@
+		Name:            "debian",
+		Version:         d.Version,
+		Description:     description,
+		ExampleUserData: *d.ExampleUserData,
```
L41 is duplicted by the block starting at L44. L41 should be dropped.

jcanocan replied:
Done.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,117 @@
+	if strings.HasPrefix(d.Version, "11") || strings.HasSuffix(d.Version, "12") {
+		baseURL = fmt.Sprintf(baseURLFmt, d.VersionName)
+		qCowImageName = fmt.Sprintf(baseQcow2Fmt, d.Version, architecture.GetImageArchitecture(d.Arch))
+	} else if strings.HasPrefix(d.Version, "13") {
```
IMO we should not ship 13 yet. Saves us this workaround?

jcanocan replied:
Besides the fact that 13 is not "stable" yet, we still want to have it available.

0xFelix replied:
ACK

dominikholler replied:
Agree, it might be a better next step to think about garbage collection old containerdisk, before adding a daily one.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,117 @@
+		return nil, fmt.Errorf("error reading the debian checksum file: %v", err)
+	}
+
+	if checksum, ok := sums[qCowImageName]; ok {
```
We should add an additional tag similar to centosstream. The images have a timestamp like `20210814-734`.

jcanocan replied:
Done.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,117 @@
+
+func (d *debian) Tests() []api.ArtifactTest {
+	return []api.ArtifactTest{
+		tests.SSH,
```
Do images have a guest agent?

jcanocan replied:
No. I does not.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+const (
+	baseURLFmt   = "https://cloud.debian.org/images/cloud/%s/latest/"
```
We should fetch the versioned images instead of `latest` to have this information available in tags. Otherwise the image will never get rebuilt once it was built.

jcanocan replied:
Added additional tag.

0xFelix commented on cmd/medius/images/push.go:
```diff
@@ -200,9 +200,11 @@ func (b *buildAndPublish) getArtifact(artifactInfo *api.ArtifactDetails) (string
-	checksum := artifactReader.Checksum()
-	if checksum != artifactInfo.SHA256Sum {
-		return "", fmt.Errorf("expected checksum %q but got %q", artifactInfo.SHA256Sum, checksum)
+	checksum256, checksum512 := artifactReader.Checksum()
```
Can you put the checksum type into `api.ArtifactDetails` instead and pass it to `b.Getter.GetWithChecksumAndContext`. The checksum returned by `artifactReader.Checksum()` could then simply be compared to `artifactInfo.Checksum`.

jcanocan replied:
I'm not sure If I'm following. The idea is to compare the checksum present in "api.ArtifactDetails", which is retrieved from ".json" in the repo, compute our own and check if they match.

0xFelix replied:
I'm asking you not to return two possible checksums from `artifactReader.Checksum()`. If you put the expected type of the checksum into `api.ArtifactDetails`, then `artifactReader.Checksum()` can return only a single checksum in the expected format.

jcanocan replied:
Ah I see. Thanks for the clarification. Done.

jcanocan commented:
v2:
- Changed the file from where the SHA checksums are obtained from the "SUM512SUMS" to individual ".json" images files.
- Added an additional tag indicating the repository build version. It will help to troubleshoot images.

jcanocan commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,171 @@
+	return metadata
+}
+
+func decodeSHA(base64Checksum string) (string, error) {
```
Done.

jcanocan commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,171 @@
+		return []string{}, "", fmt.Errorf("build debian data not found")
+	}
+
+	// Usually the qcow is on the 3rd item. However, let's be sure that it is the qcow image digest.
```
Done.

jcanocan commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,171 @@
+
+	// Usually the qcow is on the 3rd item. However, let's be sure that it is the qcow image digest.
+	for _, item := range buildData.Items {
+		if item.Metadata.Labels.ImageFormat != "" && item.Metadata.Labels.ImageFormat == "qcow2" {
```
Done.

0xFelix commented on pkg/http/http.go:
```diff
@@ -65,19 +71,34 @@ func (h *HTTPGetter) GetWithChecksumAndContext(ctx context.Context, fileURL stri
+func newReadCloserWithChecksum(body io.ReadCloser, checksumDigest api.ChecksumFormat) (readCloser *readCloserWithChecksum, err error) {
+	var checksumHash hash.Hash
+
+	switch checksumDigest {
+	case api.Sha256:
+		checksumHash = sha256.New()
+	case api.Sha512:
+		checksumHash = sha512.New()
+	default:
+		return nil, fmt.Errorf("unknown checksum digest function %v", checksumDigest)
+	}
+
```
Instead of implementing this as a switch, can't you simply pass `sha256.New` or `sha512.New` as arguments to this func and call the passed func here?

jcanocan replied:
I've directly created the checksum hasher object in the api.ArtifactDetails. In this way, it can be used directly inside this function. This also allows us to easily use any hasher.

jcanocan commented on pkg/http/http.go:
```diff
@@ -47,10 +50,13 @@ func (h *HTTPGetter) GetAllWithContext(ctx context.Context, fileURL string) ([]b
 
 func (h *HTTPGetter) GetWithChecksum(fileURL string) (ReadCloserWithChecksum, error) {
-	return h.GetWithChecksumAndContext(context.Background(), fileURL)
+	return h.GetWithChecksumAndContext(context.Background(), fileURL, "")
```
Ups, sorry. Forgot about update it. Done.

0xFelix commented on artifacts/centosstream/centos-stream.go:
```diff
@@ -91,7 +92,8 @@ func (c *centos) Inspect() (*api.ArtifactDetails, error) {
 		return &api.ArtifactDetails{
-			SHA256Sum:            checksum,
+			Checksum:             checksum,
+			ChecksumHasher:       sha256.New(),
```
```suggestion
			ChecksumHash:       sha256.New,
```

Can you do it like this?

Saves passing around the object, and is more idiomatic.

See e.g. https://pkg.go.dev/crypto/pbkdf2#Key

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,163 @@
+
+	raw, err := d.getter.GetAll(jsonURL)
+	if err != nil {
+		return []string{}, "", fmt.Errorf("error downloading debian json file: %v", err)
```
```suggestion
		return nil, "", fmt.Errorf("error downloading debian json file: %v", err)
```

Same for others below.

jcanocan replied:
Done.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		return []string{}, "", fmt.Errorf("error downloading debian json file: %v", err)
+	}
+	err = json.Unmarshal(raw, &buildData)
+	if err != nil {
```
Inline err?

jcanocan replied:
Done.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,163 @@
+		}
+	}
+
+	return additionalTags, checksum, fmt.Errorf("error locating the image information")
```
```suggestion
	return nil, "", fmt.Errorf("error locating the image information")
```

jcanocan replied:
Done.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+func (d *debian) getBuildData(jsonURL string) (additionalTags []string, checksum string, err error) {
+	var buildData BuildData
```
Move to L94?

jcanocan replied:
Done.

0xFelix commented on artifacts/debian/debian.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+func (d *debian) Inspect() (*api.ArtifactDetails, error) {
+	if !strings.HasPrefix(d.Version, "11") && !strings.HasSuffix(d.Version, "12") {
```
Why are you mixing usage of `HasPrefix` and `HasSuffix`?

jcanocan replied:
Silly copy&paste mistake, `HasPrefix` is the right one.

0xFelix commented on cmd/medius/images/push.go:
```diff
@@ -300,7 +300,7 @@ func (b *buildAndPublish) rebuildNeeded(entry *common.Entry) (bool, error) {
 		if err != nil {
 			return false, err
 		}
-		if imageSha != artifactInfo.SHA256Sum {
```
Maybe in a follow up, but `imageSha` and `b.getImageSha` should be adapted too.

jcanocan replied:
I've changed it in this PR. I think it is related with this.

0xFelix commented on cmd/medius/images/push.go:
```diff
@@ -155,14 +155,14 @@ func (b *buildAndPublish) Do(entry *common.Entry, timestamp time.Time) ([]string
 	} else {
 		b.Log.Infof("Latest containerdisk checksum: %q", imageInfo.Labels[build.LabelShaSum])
-		imageSha = imageInfo.Labels[build.LabelShaSum]
+		imageChecksum = imageInfo.Labels[build.LabelShaSum]
```
`build.LabelShaSum` can be adapted in a follow up. But not right now.

0xFelix commented on pkg/http/http.go:
```diff
@@ -13,8 +12,8 @@ import (
 	GetAllWithContext(ctx context.Context, fileURL string) ([]byte, error)
-	GetWithChecksum(fileURL string) (ReadCloserWithChecksum, error)
-	GetWithChecksumAndContext(ctx context.Context, fileURL string) (ReadCloserWithChecksum, error)
+	GetWithChecksum(fileURL string, ChecksumHasher hash.Hash) (ReadCloserWithChecksum, error)
```
Not sure an interface needs to give names to args at all, but if then I'd prefer it starting with a lowercase letter.

```suggestion
	GetWithChecksum(fileURL string, checksumHasher hash.Hash) (ReadCloserWithChecksum, error)
```

jcanocan replied:
Done.

0xFelix commented on cmd/medius/images/push.go:
```diff
@@ -186,7 +186,7 @@ func (b *buildAndPublish) handleMetadataError(imageName string, err error) error
 
 func (b *buildAndPublish) getArtifact(artifactInfo *api.ArtifactDetails) (string, error) {
-	artifactReader, err := b.Getter.GetWithChecksumAndContext(b.Ctx, artifactInfo.DownloadURL)
+	artifactReader, err := b.Getter.GetWithChecksumAndContext(b.Ctx, artifactInfo.DownloadURL, artifactInfo.ChecksumHash())
```
Why not pass the `func() hash.Hash` down to where it is actually being used?

jcanocan replied:
Done.

0xFelix reviewed (commented):
/approve

0xFelix commented on cmd/medius/common/registry.go:
```diff
@@ -83,12 +85,27 @@ var staticRegistry = []Entry{
+			debian.New("11", "bullseye", "x86_64", &docs.UserData{Username: "debian"}, nil),
+			debian.New("11", "bullseye", "aarch64", &docs.UserData{Username: "debian"}, nil),
+		},
+		UseForDocs: false,
```
Drop this line, it is the default

jcanocan replied:
Done.
